### PR TITLE
Fix metric name encode

### PR DESCRIPTION
**Pull request by bbranquinho**
### **Description**

There is an issue to publish metrics when the metric name contains special characters; this PR is to fix that issue. The best way to fix that is by using URL encoding. The metric name should use URL encoded characters, otherwise, graphite does not show the metric. Graphite shows the error "No Data" if the metric name contains special characters.

### **Error**

The following figure shows the error.

Screen Shot 2022-01-10 at 5 33 08 AM

Log: _22/01/10 05:33:30 TRACE jmx2graphite.MetricsPipeline: MetricValue{name='com.zaxxer.hikari.type_Pool-(HikariCP-writer-2).TotalConnections', value=5, timestampSeconds=1641772800}_

An easy way to test graphite is by executing the command ``echo "local.jmx.my-project.com.zaxxer.hikari.type_Pool-(HikariCP-reader-1).TotalConnections 5 `date +%s`" | nc ${GRAPHITE_SERVER} ${PORT}``

### **Fix**

Using the URLEncoder (**https://docs.oracle.com/javase/1.5.0/docs/api/java/net/URLEncoder.html**) the error is fixed.

Screen Shot 2022-01-10 at 5 20 18 AM

Log: _22/01/10 05:37:36 TRACE jmx2graphite.MetricsPipeline: MetricValue{name='com.zaxxer.hikari.type_Pool-%28HikariCP-writer-2%29.ActiveConnections', value=0, timestampSeconds=1641772800}_

Command to test graphite ``echo "local.jmx.my-project.com.zaxxer.hikari.type_Pool-%28HikariCP-reader-1%29.TotalConnections 5 `date +%s`" | nc ${GRAPHITE_SERVER} ${PORT}``

### **Additional Information**

The following PRs show the same fix applied to other projects:

https://github.com/grafana/grafana/pull/401
https://github.com/graphite-project/graphite-web/pull/1663

### CODE REVIEW

File: src/main/java/io/logz/jmx2graphite/GraphiteClient.java [hunk 1]
```diff
@@ -85,7 +89,13 @@ public static String sanitizeMetricName(String s, boolean keepDot) {
                 sb.append(c);
```
Comment (bbranquinho) [inline src/main/java/io/logz/jmx2graphite/GraphiteClient.java:89]: I could add conditions to handle parentheses () and pipes |, however, it won't solve all cases.

File: src/main/java/io/logz/jmx2graphite/GraphiteClient.java [hunk 2]
```diff
@@ -85,7 +89,13 @@ public static String sanitizeMetricName(String s, boolean keepDot) {
...
             }
         }
-        return sb.toString();
+
+        try {
+            return URLEncoder.encode(sb.toString(), ENCODING);
+        } catch (UnsupportedEncodingException e) {
+            logger.error("Unsupported encoding {}.", ENCODING, e);
```
Comment (asafm) [inline src/main/java/io/logz/jmx2graphite/GraphiteClient.java]: Let's help us solve this issue by emitting the metric name we failed to encode

**Review by asafm — CHANGES_REQUESTED** [review]
Thanks a lot for this PR - it does look needed indeed. 
Can you please a test for this - you have TestGraphiteClient (You may want to rename it to GraphiteClientTest while you're at it)

File: src/main/java/io/logz/jmx2graphite/GraphiteClient.java (hunk 2 above)
Comment (bbranquinho) [reply src/main/java/io/logz/jmx2graphite/GraphiteClient.java]: I've renamed the test class, added a unit test for the sanitizeMethodName, and fixed the error message.

**Review by asafm — APPROVED** [review]

**Comment by asafm** [comment]
@bbranquinho Something is wrong with authentication in the GitHub Actions - I've reached out to GitHub support - it will take a few days based on their previous response rate

**Pull request merged**